javalib/src/main/scala/java/io/BufferedInputStream.scala

```diff
@@ -1,129 +1,247 @@
+   * The mark is not invalidated in this method: Per spec there is no requirement to invalidate mark
+   * *exactly* when pos - markpos exceeds marklimit. This is "generous" and migh permit a reset
+   * beyond marklimit.
+   */
+  private def fillMarkedBuffer(sourceBuffer: Array[Byte],
+                               source: InputStream): Option[Array[Byte]] =
+    if (count < sourceBuffer.length) {
+      fillMarkedBufferUnsafe(sourceBuffer, source)
+    } else {
+      val newBuffer = new Array[Byte](sourceBuffer.length * 2)
+      sourceBuffer.copyToArray(newBuffer)
+      buf = newBuffer
+      fillMarkedBufferUnsafe(newBuffer, source)
+    }
 
-    if (prepareRead()) {
-      val res = buf(pos).toInt
-      pos += 1
-      res
-    } else -1
+  private def fillMarkedBufferUnsafe(
+      sourceBuffer: Array[Byte],
+      source: InputStream): Option[Array[Byte]] = {
+    val bytesRead =
+      source.read(sourceBuffer, count, sourceBuffer.length - count)
+    if (bytesRead == -1)
+      None
+    else {
+      count += bytesRead
+      Option(sourceBuffer)
+    }
   }
 
-  override def read(b: Array[Byte], off: Int, len: Int): Int = {
-    ensureOpen()
+  // can block
+  // returns -1 on end of stream
+  // otherwise returns next byte of data
+  // or throws IOException
+  override def read(): Int = (Option(buf), Option(in)) match {
+    case (Some(sourceBuffer), Some(source)) =>
+      synchronized {
+        if (pos < count) {
+          val res = sourceBuffer(pos).toInt
+          pos += 1
+          res
+        } else {
+          fillBuffer(sourceBuffer, source) match {
+            case None => -1
+            case Some(nextBuffer) => {
+              val res = nextBuffer(pos).toInt
+              pos += 1
+              res
+            }
+          }
+        }
+      }
+    case _ => throw new IOException("read requested on closed or null stream")
+  }
 
-    if (off < 0 || len < 0 || len > b.length - off)
-      throw new IndexOutOfBoundsException
+  override def read(b: Array[Byte], off: Int, len: Int): Int =
+    (Option(buf), Option(in)) match {
+      case (Some(sourceBuffer), Some(source)) => {
+        if (off < 0 || len < 0 || len > b.length - off)
+          throw new IndexOutOfBoundsException
+
+        if (len == 0) 0
+        else
+          synchronized {
+            unsafeRead(b, off, len, sourceBuffer, source)
+          }
+      }
+      case _ => throw new IOException("read(array, off, len) on closed stream")
+    }
 
-    if (len == 0) 0
-    else if (prepareRead()) {
-      val count = Math.min(len, end - pos)
-      System.arraycopy(this.buf, pos, b, off, count)
-      pos += count
-      count
-    } else -1
-  }
+  private def unsafeRead(b: Array[Byte],
+                         off: Int,
+                         len: Int,
+                         sourceBuffer: Array[Byte],
+                         source: InputStream): Int =
+    unsafeRead(b, off, len, sourceBuffer, source, 0)
+
+  @annotation.tailrec
```

Comment:
I'm really not a fan of supposedly-tailrec functions that actually mutate outside state (`pos`, the content of `b`, the call to `fillBuffer`, etc.) in addition to changing the value of some accumulators (`bytesRead`). It means that when reading this, I have to reason about the explicit state *plus* the implicit state. That's too much for my brain to handle.

Please use a good old loop with a `var bytesRead` instead.